File: api/src/atlas_api/driving.py
"""Stateless, coaching-only driving context from coarse route points."""

from __future__ import annotations

import hashlib
import json
from functools import lru_cache
from pathlib import Path
from typing import Any, Iterable

FIXTURE_PATH = Path(__file__).resolve().parents[2] / "fixtures" / "driving_context_demo.json"
COACHING_LABEL = (
    "Coaching only. This result is not underwriting, does not change a quote or premium, "
    "and is not shared with an insurer."
)


@lru_cache(maxsize=1)
def _fixture() -> dict[str, Any]:
    return json.loads(FIXTURE_PATH.read_text())


def _is_coarse(value: float) -> bool:
    return abs(value - round(value, 3)) < 1e-9


def _zone_for(lat: float, lng: float) -> dict[str, Any] | None:
    for zone in _fixture()["zones"]:
        bounds = zone["bounds"]
        if bounds["south"] <= lat <= bounds["north"] and bounds["west"] <= lng <= bounds["east"]:
            return zone
    return None


def assess_drive_context(
    *,
    points: Iterable[tuple[float, float]],
    distance_km: float,
    speeding_events: int,
    hard_brake_events: int,
) -> dict[str, Any]:
    """Assess aggregate driver events without storing or returning route coordinates."""
    route = [(float(lat), float(lng)) for lat, lng in points]
    if not 2 <= len(route) <= 50:
        raise ValueError("provide between 2 and 50 coarse route points")
    if not 0.5 <= distance_km <= 500:
        raise ValueError("distance_km must be between 0.5 and 500")
    if speeding_events < 0 or hard_brake_events < 0:
        raise ValueError("driver event counts must be zero or greater")
    for lat, lng in route:
        if not -90 <= lat <= 90 or not -180 <= lng <= 180:
            raise ValueError("route point is outside valid latitude or longitude bounds")
        if not _is_coarse(lat) or not _is_coarse(lng):
            raise ValueError("route points must be rounded to at most 3 decimal places")

    context_counts: dict[str, tuple[dict[str, Any], int]] = {}
    for lat, lng in route:
        zone = _zone_for(lat, lng)
        if zone is None:
            outside_zone = {
                "id": "outside_demo_context",
                "label": "Outside bundled Toronto context examples",
                "contextType": "outside_demo_context",
                "areaLabel": "Unclassified coarse route area",
                "contextScore": _fixture()["scoring"]["outsideDemoContextScore"],
                "source": "No matching bundled synthetic route context",
            }
            current = context_counts.get(outside_zone["id"], (outside_zone, 0))
            context_counts[outside_zone["id"]] = (outside_zone, current[1] + 1)
        else:
            current = context_counts.get(zone["id"], (zone, 0))
            context_counts[zone["id"]] = (zone, current[1] + 1)

    scoring = _fixture()["scoring"]
    speeding_rate = speeding_events / distance_km * 100
    hard_brake_rate = hard_brake_events / distance_km * 100
    speeding_penalty = min(
        int(scoring["speedingPenaltyCap"]),
        round(speeding_rate * float(scoring["speedingPenaltyPerEventPer100Km"])),
    )
    hard_brake_penalty = min(
        int(scoring["hardBrakePenaltyCap"]),
        round(hard_brake_rate * float(scoring["hardBrakePenaltyPerEventPer100Km"])),
    )
    behavior_score = max(0, int(scoring["startingScore"]) - speeding_penalty - hard_brake_penalty)
    behavior_weight = float(scoring["behaviorWeight"])
    route_context_weight = float(scoring["routeContextWeight"])

    route_factors = []
    route_context_score = 0.0
    for zone, count in context_counts.values():
        point_share = count / len(route)
        context_contribution = float(zone["contextScore"]) * point_share
        route_context_score += context_contribution
        route_factors.append(
            {
                "key": zone["id"],
                "label": zone["label"],
                "contextType": zone["contextType"],
                "areaLabel": zone["areaLabel"],
                "coarsePointsMatched": count,
                "pointShare": round(point_share, 4),
                "contextScore": zone["contextScore"],
                "routeContextContribution": round(context_contribution, 2),
                "compositeContribution": round(context_contribution * route_context_weight, 2),
                "source": zone["source"],
                "provenance": "bundled_synthetic_route_context",
                "meaning": (
                    "Estimates route exposure and attention demand only. A lower context score means "
                    "more attention may be useful; it does not judge the driver or affect insurance pricing."
                ),
            }
        )
    route_context_score = round(route_context_score, 2)
    behavior_contribution = behavior_score * behavior_weight
    route_context_contribution = route_context_score * route_context_weight
    composite_score = int(behavior_contribution + route_context_contribution + 0.5)
    band = "steady" if composite_score >= 85 else "watch" if composite_score >= 65 else "focus"

    factors = [
        {
            "key": "speeding",
            "label": "Speeding events",
            "observed": speeding_events,
            "ratePer100Km": round(speeding_rate, 2),
            "effectPoints": -speeding_penalty,
            "source": (
                f"Caller-supplied aggregate count and distance; {scoring['speedingPenaltyPerEventPer100Km']} "
                f"points per event per 100 km, capped at {scoring['speedingPenaltyCap']}"
            ),
            "provenance": "request_aggregate_and_bundled_demo_formula",
        },
        {
            "key": "hard_brake",
            "label": "Hard-brake events",
            "observed": hard_brake_events,
            "ratePer100Km": round(hard_brake_rate, 2),
            "effectPoints": -hard_brake_penalty,
            "source": (
                f"Caller-supplied aggregate count and distance; {scoring['hardBrakePenaltyPerEventPer100Km']} "
                f"points per event per 100 km, capped at {scoring['hardBrakePenaltyCap']}"
            ),
            "provenance": "request_aggregate_and_bundled_demo_formula",
        },
    ]
    tips = []
    if speeding_events:
        tips.append("Leave more time for the trip so posted speed changes do not require catching up.")
    if hard_brake_events:
        tips.append("Increase following distance and scan farther ahead for stops.")
    if not tips:
        tips.append("Keep the same pace and following distance on the next drive.")

    assessment_key = json.dumps(
        {
            "contexts": sorted(
                (factor["contextType"], factor["coarsePointsMatched"])
                for factor in route_factors
            ),
            "distanceKm": round(distance_km, 1),
            "speedingEvents": speeding_events,
            "hardBrakeEvents": hard_brake_events,
        },
        sort_keys=True,
    )
    return {
        "assessmentId": "DRV-" + hashlib.sha256(assessment_key.encode()).hexdigest()[:8],
        "coachingOnly": True,
        "affectsQuote": False,
        "affectsPremium": False,
        "score": composite_score,
        "behaviorScore": behavior_score,
        "routeContextScore": route_context_score,
        "composite": {
            "behaviorWeight": behavior_weight,
            "routeContextWeight": route_context_weight,
            "behaviorContribution": round(behavior_contribution, 2),
            "routeContextContribution": round(route_context_contribution, 2),
            "formula": (
                f"behaviorScore × {behavior_weight:.2f} + "
                f"routeContextScore × {route_context_weight:.2f}"
            ),
        },
        "band": band,
        "factors": factors,
        "routeFactors": route_factors,
        "tips": tips,
        "routeContext": {
            "labels": [factor["label"] for factor in route_factors],
            "types": [factor["contextType"] for factor in route_factors],
            "coarsePointsReceived": len(route),
            "rawCoordinatesReturned": False,
            "meaning": (
                "The route-context score estimates exposure and attention needs. "
                "It does not indicate whether the driver behaved well or poorly."
            ),
        },
        "privacy": {
            "stored": False,
            "identityInputsUsed": False,
            "rawCoordinateRetention": "none",
            "note": "Coordinates are classified during this request and then discarded.",
        },
        "source": _fixture()["label"],
        "label": COACHING_LABEL,
    }

File: app/lib/driving.ts
export interface DriveAssessment {
  assessmentId: string;
  coachingOnly: true;
  affectsQuote: false;
  affectsPremium: false;
  score: number;
  behaviorScore: number;
  routeContextScore: number;
  composite: { behaviorWeight: number; routeContextWeight: number; behaviorContribution: number; routeContextContribution: number; formula: string };
  band: 'steady' | 'watch' | 'focus';
  factors: Array<{ key: 'speeding' | 'hard_brake'; label: string; observed: number; ratePer100Km: number; effectPoints: number; source: string }>;
  routeFactors: Array<{ key: string; label: string; areaLabel: string; contextScore: number; meaning: string }>;
  tips: string[];
  routeContext: { labels: string[]; types: string[]; coarsePointsReceived: number; rawCoordinatesReturned: false; meaning: string };
  privacy: { stored: false; rawCoordinateRetention: 'none'; note: string };
  label: string;
}

export type DriveAssessmentResult = { assessment: DriveAssessment; source: 'shared-api' | 'bundled-demo' };

export type DriveSessionSummary = {
  score: number;
  behaviorScore: number;
  routeContextScore: number;
  band: DriveAssessment['band'];
  distanceKm: number;
  durationSeconds: number;
  currentSpeedKmh: number;
  maxSpeedKmh: number;
  speedingEvents: number;
  hardBrakeEvents: number;
  area: string;
  updatedAt: number;
  source: DriveAssessmentResult['source'];
};

export type DriveAssessmentInput = {
  points?: Array<{ lat: number; lng: number }>;
  distanceKm?: number;
  speedingEvents?: number;
  hardBrakeEvents?: number;
};

const API = process.env.EXPO_PUBLIC_API_URL;
const SYNTHETIC_ROUTE = {
  points: [
    { lat: 43.639, lng: -79.443 },
    { lat: 43.642, lng: -79.408 },
    { lat: 43.650, lng: -79.381 },
  ],
  distanceKm: 9.4,
  speedingEvents: 0,
  hardBrakeEvents: 1,
};

const FALLBACK: DriveAssessment = {
  assessmentId: 'DRV-DEMO-TORONTO',
  coachingOnly: true,
  affectsQuote: false,
  affectsPremium: false,
  score: 87,
  behaviorScore: 91,
  routeContextScore: 73.33,
  composite: { behaviorWeight: 0.75, routeContextWeight: 0.25, behaviorContribution: 68.25, routeContextContribution: 18.33, formula: 'behaviorScore × 0.75 + routeContextScore × 0.25' },
  band: 'steady',
  factors: [
    { key: 'speeding', label: 'Speeding events', observed: 0, ratePer100Km: 0, effectPoints: 0, source: 'Bundled synthetic route' },
    { key: 'hard_brake', label: 'Hard-brake events', observed: 1, ratePer100Km: 10.64, effectPoints: -9, source: 'Bundled synthetic route and demo formula' },
  ],
  routeFactors: [
    { key: 'outside_demo_context', label: 'Outside bundled Toronto context examples', areaLabel: 'Unclassified coarse route area', contextScore: 80, meaning: 'No matching bundled context.' },
    { key: 'dense_intersection_demo', label: 'Dense intersection and building corridor', areaLabel: 'Downtown Toronto demo area', contextScore: 68, meaning: 'Higher attention demand in the synthetic route context.' },
    { key: 'school_approach_demo', label: 'School approach', areaLabel: 'Downtown Toronto demo area', contextScore: 72, meaning: 'Higher attention demand in the synthetic route context.' },
  ],
  tips: ['Leave more space before dense intersections.'],
  routeContext: { labels: ['Outside bundled Toronto context examples', 'Dense intersection and building corridor', 'School approach'], types: ['outside_demo_context', 'dense_intersection_building_corridor', 'school_approach'], coarsePointsReceived: 3, rawCoordinatesReturned: false, meaning: 'Route context measures exposure and attention demand.' },
  privacy: { stored: false, rawCoordinateRetention: 'none', note: 'Synthetic route only. Raw coordinates are not retained.' },
  label: 'Coaching only. This does not change a quote or premium.',
};

function valid(value: unknown): value is DriveAssessment {
  if (!value || typeof value !== 'object') return false;
  const item = value as Partial<DriveAssessment>;
  return typeof item.assessmentId === 'string' && typeof item.score === 'number' && typeof item.behaviorScore === 'number' && typeof item.routeContextScore === 'number' && !!item.composite && Array.isArray(item.factors) && Array.isArray(item.routeFactors) && item.affectsPremium === false;
}

function fallbackAssessment(input: Required<DriveAssessmentInput>): DriveAssessment {
  const speedingPenalty = Math.min(35, input.speedingEvents * 8);
  const brakePenalty = Math.min(30, input.hardBrakeEvents * 7);
  const behaviorScore = Math.max(0, 100 - speedingPenalty - brakePenalty);
  const routeContextScore = FALLBACK.routeContextScore;
  const score = Math.round(behaviorScore * 0.75 + routeContextScore * 0.25);
  return {
    ...FALLBACK,
    assessmentId: `DRV-LOCAL-${Date.now().toString().slice(-6)}`,
    score,
    behaviorScore,
    band: score >= 85 ? 'steady' : score >= 65 ? 'watch' : 'focus',
    composite: {
      ...FALLBACK.composite,
      behaviorContribution: Number((behaviorScore * 0.75).toFixed(2)),
      routeContextContribution: Number((routeContextScore * 0.25).toFixed(2)),
    },
    factors: [
      { ...FALLBACK.factors[0], observed: input.speedingEvents, effectPoints: -speedingPenalty, source: 'Foreground GPS speed samples and bundled demo formula' },
      { ...FALLBACK.factors[1], observed: input.hardBrakeEvents, effectPoints: -brakePenalty, source: 'Foreground GPS speed changes and bundled demo formula' },
    ],
    label: 'Coaching preview from this foreground session. This does not change a quote or premium.',
  };
}

export async function assessDrive(input: DriveAssessmentInput = {}): Promise<DriveAssessmentResult> {
  const payload: Required<DriveAssessmentInput> = {
    points: input.points && input.points.length >= 2 ? input.points.slice(-50) : SYNTHETIC_ROUTE.points,
    distanceKm: Math.max(0.5, input.distanceKm ?? SYNTHETIC_ROUTE.distanceKm),
    speedingEvents: input.speedingEvents ?? SYNTHETIC_ROUTE.speedingEvents,
    hardBrakeEvents: input.hardBrakeEvents ?? SYNTHETIC_ROUTE.hardBrakeEvents,
  };
  if (API) {
    try {
      const response = await fetch(`${API}/driving/context`, {
        method: 'POST', headers: { 'content-type': 'application/json' }, body: JSON.stringify(payload), signal: AbortSignal.timeout(6000),
      });
      if (response.ok) {
        const payload: unknown = await response.json();
        if (valid(payload)) return { assessment: payload, source: 'shared-api' };
      }
    } catch {}
  }
  return { assessment: fallbackAssessment(payload), source: 'bundled-demo' };
}
